Sources/MongoSwift/ClientSession.swift

```diff
@@ -56,34 +56,58 @@ private func withSessionOpts<T>(wrapping options: ClientSessionOptions?,
 public final class ClientSession {
     /// Error thrown when an inactive session is used.
     internal static var SessionInactiveError = UserError.logicError(message: "Tried to use an inactive session")
+    /// Error thrown when a user attempts to use a session with a client it was not created from.
+    internal static var ClientMismatchError = UserError.invalidArgumentError(
+        message: "Sessions may only be used with the MongoClient used to create them")
+
+    /// Enum for tracking the state of a session.
+    internal enum State {
+    /// Indicates that the session is active. Stores a pointer to the underlying `mongoc_client_session_t` and the
+    /// source `Connection` for this session.
+    case active(session: OpaquePointer, connection: Connection)
+    /// Indicates that the session has been ended.
+    case inactive
+    }
 
-    /// Pointer to the underlying `mongoc_client_session_t`.
-    internal fileprivate(set) var _session: OpaquePointer?
+    /// Indicates the state of this session.
+    internal private(set) var state: State
 
     /// Returns whether this session has been ended or not.
-    internal var active: Bool { return self._session != nil }
+    internal var active: Bool {
+        if case .active = self.state {
```

Comment:
since `State` isn't `Equatable` you have to match it with the case syntax. 
I considered making it `Equatable` but it feels a little weird as we would have to ignore the associated values in the `active` case (or also somehow make `connection` `Equatable`) and this is the only place we'd use it